File: backend/src/main/java/com/interview/controller/NotificationController.java
package com.interview.controller;

import com.interview.dto.ApiResponse;
import com.interview.model.Notification;
import com.interview.service.NotificationService;
import com.interview.util.SecurityUtils;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST controller for notification operations
 */
@Slf4j
@RestController
@RequestMapping("/api/v1/notifications")
@RequiredArgsConstructor
@Tag(name = "Notification", description = "Notification API")
public class NotificationController {

    private final NotificationService notificationService;
    private final SecurityUtils securityUtils;

    @GetMapping
    @Operation(summary = "Get all notifications")
    public ResponseEntity<ApiResponse<List<Notification>>> getNotifications() {
        log.info("GET /api/v1/notifications");
        String userId = securityUtils.getCurrentUserId();
        List<Notification> notifications = notificationService.getUserNotifications(userId);
        return ResponseEntity.ok(ApiResponse.success(notifications));
    }

    @GetMapping("/paginated")
    @Operation(summary = "Get paginated notifications")
    public ResponseEntity<ApiResponse<Page<Notification>>> getNotificationsPaginated(Pageable pageable) {
        log.info("GET /api/v1/notifications/paginated");
        String userId = securityUtils.getCurrentUserId();
        Page<Notification> notifications = notificationService.getUserNotificationsPaginated(userId, pageable);
        return ResponseEntity.ok(ApiResponse.success(notifications));
    }

    @GetMapping("/unread")
    @Operation(summary = "Get unread notifications")
    public ResponseEntity<ApiResponse<List<Notification>>> getUnreadNotifications() {
        log.info("GET /api/v1/notifications/unread");
        String userId = securityUtils.getCurrentUserId();
        List<Notification> notifications = notificationService.getUnreadNotifications(userId);
        return ResponseEntity.ok(ApiResponse.success(notifications));
    }

    @GetMapping("/unread/count")
    @Operation(summary = "Get unread notification count")
    public ResponseEntity<ApiResponse<Long>> getUnreadCount() {
        log.info("GET /api/v1/notifications/unread/count");
        String userId = securityUtils.getCurrentUserId();
        long count = notificationService.getUnreadCount(userId);
        return ResponseEntity.ok(ApiResponse.success(count));
    }

    @PatchMapping("/{notificationId}/read")
    @Operation(summary = "Mark notification as read")
    public ResponseEntity<ApiResponse<String>> markAsRead(@PathVariable String notificationId) {
        log.info("PATCH /api/v1/notifications/{}/read", notificationId);
        notificationService.markAsRead(notificationId);
        return ResponseEntity.ok(ApiResponse.success("Notification marked as read", null));
    }

    @PatchMapping("/read-all")
    @Operation(summary = "Mark all notifications as read")
    public ResponseEntity<ApiResponse<String>> markAllAsRead() {
        log.info("PATCH /api/v1/notifications/read-all");
        String userId = securityUtils.getCurrentUserId();
        notificationService.markAllAsRead(userId);
        return ResponseEntity.ok(ApiResponse.success("All notifications marked as read", null));
    }
}

File: backend/src/controllers/notification.controller.js
const notificationService = require('../services/notification.service');
const ApiResponse = require('../utils/api-response');
const { getCurrentUserId } = require('../utils/security.utils');

const getNotifications = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const notifications = await notificationService.getUserNotifications(userId);
    return res.status(200).json(ApiResponse.success(notifications));
  } catch (error) {
    next(error);
  }
};

const getNotificationsPaginated = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const page = parseInt(req.query.page) || 0;
    const size = parseInt(req.query.size) || 10;
    const notifications = await notificationService.getUserNotificationsPaginated(userId, page, size);
    return res.status(200).json(ApiResponse.success(notifications));
  } catch (error) {
    next(error);
  }
};

const getUnreadNotifications = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const notifications = await notificationService.getUnreadNotifications(userId);
    return res.status(200).json(ApiResponse.success(notifications));
  } catch (error) {
    next(error);
  }
};

const getUnreadCount = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    const count = await notificationService.getUnreadCount(userId);
    return res.status(200).json(ApiResponse.success(count));
  } catch (error) {
    next(error);
  }
};

const markAsRead = async (req, res, next) => {
  try {
    await notificationService.markAsRead(req.params.notificationId);
    return res.status(200).json(ApiResponse.success('Notification marked as read', null));
  } catch (error) {
    next(error);
  }
};

const markAllAsRead = async (req, res, next) => {
  try {
    const userId = getCurrentUserId(req);
    await notificationService.markAllAsRead(userId);
    return res.status(200).json(ApiResponse.success('All notifications marked as read', null));
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getNotifications,
  getNotificationsPaginated,
  getUnreadNotifications,
  getUnreadCount,
  markAsRead,
  markAllAsRead
};
